File: Arithmetic/fft-poly-mult.py
# Algorithm: Fast Fourier Transform (for Polynomial Multiplication)
# Type: Divide and Conquer, Computational Algebra
# Time Complexity: O(n log n)
# Space Complexity: O(n)

import cmath

def fft(a):
    n = len(a)
    if n == 1:
        return a
    w_n = cmath.exp(2j * cmath.pi / n)
    w = 1
    a_even = fft(a[0::2])
    a_odd = fft(a[1::2])
    y = [0] * n
    for k in range(n // 2):
        t = w * a_odd[k]
        y[k] = a_even[k] + t
        y[k + n // 2] = a_even[k] - t
        w *= w_n
    return y

def ifft(a):
    n = len(a)
    a_conj = [x.conjugate() for x in a]
    y = fft(a_conj)
    return [x.conjugate() / n for x in y]

def polynomial_multiply(p, q):
    n = 1
    while n < len(p) + len(q) - 1:
        n *= 2
    p += [0] * (n - len(p))
    q += [0] * (n - len(q))
    fft_p = fft([complex(x) for x in p])
    fft_q = fft([complex(x) for x in q])
    fft_r = [fft_p[i] * fft_q[i] for i in range(n)]
    result = ifft(fft_r)
    return [round(r.real) for r in result]

File: Arithmetic/fft-poly-mult.c
// Algorithm: Fast Fourier Transform (for Polynomial Multiplication)
// Type: Divide and Conquer, Computational Algebra
// Time Complexity: O(n log n)
// Space Complexity: O(n)

// Pseudocode for FFT in C-like structure using complex numbers
// Requires a library or custom complex number handling

void fft(Complex *a, int n, int invert) {
    if (n == 1) return;

    Complex a_even[n/2], a_odd[n/2];
    for (int i = 0; i < n / 2; ++i) {
        a_even[i] = a[2*i];
        a_odd[i] = a[2*i+1];
    }

    fft(a_even, n/2, invert);
    fft(a_odd, n/2, invert);

    double angle = 2 * M_PI / n * (invert ? -1 : 1);
    Complex w = {1, 0}, wn = {cos(angle), sin(angle)};
    for (int i = 0; i < n / 2; ++i) {
        Complex t = multiply(w, a_odd[i]);
        a[i] = add(a_even[i], t);
        a[i + n/2] = subtract(a_even[i], t);
        w = multiply(w, wn);
    }

    if (invert)
        for (int i = 0; i < n; ++i)
            a[i] = divide_by_scalar(a[i], 2);
}

